File: ui/src/checkers/networking/internet_connectivity.ts
/**
 * Checks internet connectivity test results.
 */

// Ours
import { BaseChecker, type CheckResult } from '../base_checker'
import type { DataLookup } from '../data_lookup'

export class InternetConnectivityChecker extends BaseChecker {
    readonly path: string = 'networking/internet_connectivity'
    readonly name: string = 'Internet connectivity'
    readonly category: string = 'Networking / Calling'
    readonly interest = 'Health' as const
    readonly explanation: string = 'When internet access is declared, verifies actual connectivity via DNS resolution, TCP, and HTTP tests.'

    check(data: DataLookup): CheckResult {
        if (data.config && !data.config.options.has_internet) {
            return { status: 'not_applicable', status_reason: 'Internet access not declared.' }
        }

        const point = data.get('preflight/internet_connectivity')
        if (!point?.value) return { status: 'gather_failure', status_reason: 'Internet connectivity test data not collected.' }

        let parsed: Record<string, unknown> | null = null
        try { parsed = JSON.parse(String(point.value)) } catch { /* ignore */ }
        if (!parsed) return { status: 'gather_failure', status_reason: 'Could not parse data.' }

        const all_ok: boolean = parsed.all_ok as boolean ?? false
        const details: string = (parsed.details as string) ?? ''

        if (all_ok) {
            return { status: 'healthy', status_reason: `Internet connectivity verified: ${details}`, display_value: 'connected', raw_output: point.raw_output }
        }

        return {
            status: 'unhealthy',
            status_reason: `Internet access declared but connectivity test **failed**: ${details}. Targets requiring internet (push notifications, AWS endpoints) will fail.`,
            display_value: 'failed',
            raw_output: point.raw_output,
        }
    }
}

export default InternetConnectivityChecker

File: src/targets/preflight/internet_connectivity.py
"""Tests actual internet connectivity when the deployment declares internet access.

Performs a three-step connectivity test:
1. DNS resolution of dns.google (proves DNS works for public names)
2. TCP connect to 8.8.8.8:53 (proves IP-level internet routing)
3. HTTP GET to http://www.gstatic.com/generate_204 (proves HTTP works)

If the user says they have internet but this test fails, that's a problem.
"""

from __future__ import annotations

# External
import json
import socket
import urllib.request
import urllib.error

# Ours
from src.lib.base_target import BaseTarget, NotApplicableError


class InternetConnectivity(BaseTarget):
    """Test internet connectivity from wherever the runner executes.

    Only runs when has_internet is true in the deployment configuration.
    """

    @property
    def description(self) -> str:
        """What this target checks."""
        return "Internet connectivity test"

    @property
    def explanation(self) -> str:
        """Why this matters."""
        return (
            "If the deployment declares internet access, we verify it actually works. "
            "Internet is needed for real AWS push notifications, AWS SES email, "
            "pulling container images, and some DNS operations."
        )

    @property
    def unit(self) -> str:
        """No unit — returns structured JSON string."""
        return ""

    def collect(self) -> str:
        """Perform a three-step internet connectivity test.

        Returns:
            JSON string with per-step results and overall status.

        Raises:
            NotApplicableError: If internet access is not declared.
        """
        if not self.config.options.has_internet:
            raise NotApplicableError("Internet access is not declared in the deployment configuration")

        self.terminal.step("Testing internet connectivity...")

        dns_ok: bool = False
        tcp_ok: bool = False
        http_ok: bool = False
        details: list[str] = []

        # Step 1: DNS resolution of a well-known public hostname
        self.terminal.step("  DNS: resolving dns.google...")
        try:
            socket.getaddrinfo("dns.google", 53, socket.AF_INET, socket.SOCK_STREAM)
            dns_ok = True
            details.append("DNS: ok")
        except (socket.gaierror, OSError) as e:
            details.append(f"DNS: failed ({e})")

        # Step 2: TCP connect to a well-known public IP
        self.terminal.step("  TCP: connecting to 8.8.8.8:53...")
        try:
            with socket.create_connection(("8.8.8.8", 53), timeout=10):
                tcp_ok = True
                details.append("TCP: ok")
        except (socket.timeout, OSError) as e:
            details.append(f"TCP: failed ({e})")

        # Step 3: HTTP GET to a lightweight endpoint
        self.terminal.step("  HTTP: fetching http://www.gstatic.com/generate_204...")
        try:
            req = urllib.request.Request(
                "http://www.gstatic.com/generate_204",
                method="GET",
            )
            with urllib.request.urlopen(req, timeout=10) as resp:
                if resp.status == 204:
                    http_ok = True
                    details.append("HTTP: ok (204)")
                else:
                    details.append(f"HTTP: unexpected status {resp.status}")
        except (urllib.error.URLError, OSError) as e:
            details.append(f"HTTP: failed ({e})")

        all_ok: bool = dns_ok and tcp_ok and http_ok

        result: dict[str, object] = {
            "dns_ok": dns_ok,
            "tcp_ok": tcp_ok,
            "http_ok": http_ok,
            "all_ok": all_ok,
            "details": ", ".join(details),
        }

        if all_ok:
            self._health_info = "Internet connectivity: all tests passed"
        else:
            self._health_info = f"Internet connectivity: {', '.join(details)}"

        return json.dumps(result)
